File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> bracks = new Stack<>();
        Map<Character,Character> closers = new HashMap<>(3);

        closers.put(')','(');
        closers.put(']','[');
        closers.put('}','{');
        for(int i = 0; i < s.length(); i++){
            char c = s.charAt(i);
            if(closers.containsKey(c)) {
                if(!bracks.isEmpty() && closers.get(c).equals(bracks.peek())) {
                    bracks.pop();
                } else {
                    return false;
                }
            }else{
                bracks.push(c);
            }
        }
        return bracks.isEmpty();


    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        # stack = {  "{":"}","["::"]","(":")"}
        # count1, count2, count3 = 0,0,0
        # for c in s:
        #     if c == "{":
        #         count1+=1
        #     elif c == "[":
        #         count2+=1
        #     elif c == "(":
        #         count3+=1
        # for c in s:
        #     if c == "}" and count1>=1:
        #         count1-=1
        #     elif c == "]" and count2>=1:
        #         count2-=1
        #     elif c == ")" and count3>=1:
        #         count3-=1
        # if count1==0 and count2==0 and count3==0:
        #     return True
        # return False

        hashMap = {  "}":"{","]":"[",")":"("}
        res =[]
        for c in s:
            if c not in hashMap:
                res.append(c)
            else:
                if not res:
                    return False
                else:
                    out = res.pop()
                    if out != hashMap[c]:
                        return False
        return not res




                
